go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,271 @@
+	//TODO: if the old and new resolutions are the same should not update the dnsObject
+	e.dnsEntries[dnsNamespace.dnsName].dnsResolves = ips
+
+	//update the dnsObject
+	dnsObject, err := e.wf.GetDNSObject(e.nodeName)
+	existed := true
+	if errors.IsNotFound(err) {
+		// the object *MOST LIKELY* has not been created yet
+		dnsObject = &dnsobjectapi.DNSObject{
+			ObjectMeta: metav1.ObjectMeta{Name: e.nodeName},
+			Spec:       dnsobjectapi.DNSObjectSpec{},
+		}
+		existed = false
+	} else if err != nil {
+		return err
+	}
+	var ipStrings []string
+	for _, ip := range e.dnsEntries[dnsNamespace.dnsName].dnsResolves {
+		ipStrings = append(ipStrings, ip.String())
+	}
+
+	if dnsObject.Spec.DNSObjectEntries == nil {
+		dnsObject.Spec.DNSObjectEntries = make(map[string]dnsobject.DNSObjectEntry)
+	}
+
+	dnsObjectEntry := dnsObject.Spec.DNSObjectEntries[dnsNamespace.dnsName]
+	dnsObjectEntry.IPAddresses = ipStrings
+	dnsObject.Spec.DNSObjectEntries[dnsNamespace.dnsName] = dnsObjectEntry
+
+	if existed {
+		err = e.k.UpdateDNSObject(dnsObject)
+	} else {
+		_, err = e.k.CreateDNSObject(dnsObject)
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	return e.dns.Update(dns)
+}
+
+// Run spawns a goroutine that handles updates to the dns entries for dnsNames used in
+// EgressFirewalls. The loop runs after receiving one of two signals
+// 1. a new dnsName has been added and a signal is sent to add the new DNS name, if an
+//    EgressFirewall uses a DNS name already added by another egressFirewall the previous
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsNamespace dnsNamespace
+	var ttl time.Time
+	var timeSet bool
+	// initially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsNamespace := <-e.added:
+				if err := e.dns.Add(dnsNamespace.dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsNamespace); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
+				if len(dnsNamespace.dnsName) > 0 {
+					if _, err := e.Update(dnsNamespace.dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+					if err := e.updateEntryForName(dnsNamespace); err != nil {
+						utilruntime.HandleError(err)
+					}
+				}
+			case dnsName := <-e.removed:
+				// when there is a dnsName removed the query time should be updated
+				e.dns.Delete(dnsName)
+			case <-e.stopChan:
+				return
+			case <-e.controllerStop:
+				return
+			}
+
+			// before waiting on the signals get the next time this thread needs to wake up
+			ttl, dnsNamespace.dnsName, timeSet = e.dns.GetNextQueryTime()
```

Comment:
Hmm, shouldn't these be two different goroutines? One that handles new names, and one that handles periodic refreshes? Otherwise updates could be lost.

Reply:
@squeed since this is the way it used to function can I make a git issue and change this in a follow up?